### PR TITLE
#159204721 User Create, Update and Delete Articles  

### PR DESCRIPTION
**Pull request by Oluwafayokemi**
#### What does this PR do?
An authenticated user can create, update and delete an article
#### Description of Task to be completed?
- create models for an article, comment, reply, tag and rating
- An authenticated user should be able to create only the articles belonging to him
- An authenticated user should be able to update only the articles belonging to him
- An authenticated user should be able to delete only the articles belonging to him
- Generate two different unique slugs when two users create articles of the same title
- Article slugs should be automatically added to articles if they do not exist
- Add Timestamp to the articles
- Write unit test(s)
#### How should this be manually tested?
- Run npm  run dev or npm start on your terminal
##### Open postman:
- To create an article make a post request to http://localhost:3000/api/articles
- To update an article make a put request to http://localhost:3000/api/articles/:slug
- To delete an article make a delete request to http://localhost:3000/api/articles
#### What are the relevant pivotal tracker stories?
[#159204721](https://www.pivotaltracker.com/story/show/159204721)
#### Screenshots (if appropriate)

image

image
image

### CODE REVIEW

File: server/controllers/ArticleController.js [hunk 1]
```diff
@@ -1,16 +1,180 @@
...
+        message: 'Article has been created',
+        article: {
+          slug: newArticle.slug,
+          authorId: newArticle.userId,
+          categoryId: newArticle.categoryId,
+          title: newArticle.title,
+          body: newArticle.body,
+          imageUrl: newArticle.imageUrl,
+          Author: username,
+          createdAt: new Date(newArticle.createdAt).toLocaleString('en-GB', { hour12: true }),
```
Comment (madeofhuman) [inline server/controllers/ArticleController.js]: 🙌🏾🙌🏾🙌🏾Nice thinking here, @Oluwafayokemi!!! I notice, however, that you're returning both `authorId` and `Author`. I feel that is a bit redundant.

**Review by madeofhuman — CHANGES_REQUESTED** [review]
You did a really good job, Fayoke, I especially love how you converted the time to make it more human-readable in the article response. Do find time, however, to address my concern about the `aruthorId` and `Author` returned.

Furthermore, I noticed you were declaring the foreign key columns in the models. You only declare those in the migration file. The association is what is specified in both the migration file and the model file.

Overall, good work!

**Comment by Oluwafayokemi** [comment]
Thank you for the quick feedback CJ. I have implemented the feedback.

**Review by Veraclins — CHANGES_REQUESTED** [review]
@Oluwafayokemi Nice job you did here. There are just a few areas I feel you might want to look into again.
1. How are you ensuring that only a logged in user is able to perform these operations? I may be wrong but I think you should have protected the article routes with the isloggedIn middleware.
2. I see that you are duplicating a lot of codes in the response you are returning to the user. You could create a method that takes the article created or updated and return the response. This way, you will write the code just once and easily change the structure of the response while maintaining the same response.

File: server/controllers/ArticleController.js [hunk 2]
```diff
@@ -1,16 +1,177 @@
...
+    } = req.body;
+
+    Article.create({
+      slug: slug.toLowerCase(),
+      title,
+      body,
+      imageUrl,
+      categoryId,
+      userId: id,
+      createdAt: Date.now(),
```
Comment (Veraclins) [inline server/controllers/ArticleController.js]: @Oluwafayokemi, is there a reason you are adding createdAt and updatedAt in your query? Sequelize already does this. I feel you are manually redoing what is already done authomatically.

**Review by Veraclins — COMMENTED** [review]
@Oluwafayokemi, is there a reason you are adding createdAt and updatedAt in your query? Sequelize already does this. I feel you are manually redoing what is already done authomatically.

**Review by chukwuemekachm — CHANGES_REQUESTED** [review]
Well done @Oluwafayokemi on completing this task. However, a few key things to note are as follows.

- Just like @Veraclins  pointed out, you should ensure the endpoints are protected. This prevents unauthenticated and unauthorized users from performing operations on those endpoints.

- On line `95` of `ArticleController.js`, I feel sending the slug to the database is redundant since you are not updating it. I might be wrong and so I would like to learn your reasons for doing that.

- From line `68` through `170` of `ArticleController.js` I noticed that you also allowed the user to perform an update and delete operation on a resource in the database without verifying that the resource in question belonged to them. Please try to add a mechanism which verifies that the user trying to perform these operations owns the resources. This would prevent authors from updating or deleting each other's articles.

- From line `29` through `90` of `ArticleValidation` your error responses are not in line with the team's convention and API specs. The team's convention for an error response should contain an `errors` property in the response. The errors property should be an object which has properties of the illegal request payload's properties as arrays. These arrays would contain all the failing validation for their individual properties. Please kindly resolve these so that all endpoints would have a uniform/consistent error response.

Welldone.

**Comment by Oluwafayokemi** [comment]
Thank you for the feedback Chima and Clinton. I actually authenticated the
endpoints. I passed the function as a middleware to the article route.
Please check it.
Clinton, I am not sure of what you mean by duplicating code, but I will
look into it. Chima, the reason why I returned the slug was to prove that
the slug doesn't change after you update, but if the team doesn't want it
returned I will remove it. I will attend to every feedback as fast as I
can. Thank you!

File: server/controllers/ArticleController.js (hunk 2 above)
Comment (Oluwafayokemi) [reply server/controllers/ArticleController.js]: @Veraclins I have tested sequelize date and time. It's not wholly dependent. That is why I am creating mine and passing it as a query string. I was also able to manage the created at and updated at, else sequelize was going to return the same date and time for the both of them.

**Review by Veraclins — APPROVED** [review]

**Review by chukwuemekachm — APPROVED** [review]

**Review by madeofhuman — APPROVED** [review]
Good job finishing this, @Oluwafayokemi. I know it was not easy, especially with the conflicting and changing models and the merge conflicts from hell. That showed great determination. Keep it up.

**Pull request merged**